fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/error/ErrorUtils.java

```diff
@@ -0,0 +1,66 @@
+package com.github.dedis.popstellar.utility.error;
+
+import android.content.Context;
+import android.util.Log;
+import android.widget.Toast;
+
+import androidx.annotation.StringRes;
+
+import com.github.dedis.popstellar.R;
+
+import org.bouncycastle.util.Arrays;
+
+import java.util.concurrent.TimeoutException;
+
+public class ErrorUtils {
+
+  private ErrorUtils() throws IllegalAccessException {
+    throw new IllegalAccessException();
+  }
+
+  /**
+   * Log and error and show a localized {@link Toast} to the user
+   *
+   * <p>action is expected to leave a string placeholder for the error specific message at the last
+   * place
+   *
+   * <p>Ex: "An error occurred while processing %1$s : %2$s
+   *
+   * @param context to retrieve the localized messages and show the toast on
+   * @param tag used to log the error
+   * @param error to log
+   * @param action short description of what was happening when the error occurred
+   * @param actionArgs arguments to the message
+   */
+  public static void logAndShow(
+      Context context, String tag, Throwable error, @StringRes int action, String... actionArgs) {
+    String exceptionMsg = getLocalizedMessage(context, error, action, actionArgs);
+
+    Log.e(tag, exceptionMsg, error);
+    Toast.makeText(context, exceptionMsg, Toast.LENGTH_LONG).show();
+  }
+
+  private static String getLocalizedMessage(
+      Context context, Throwable error, @StringRes int action, String... actionArgs) {
+    String errorSpecificMessage = getErrorSpecificMessage(context, error);
+    actionArgs = Arrays.append(actionArgs, errorSpecificMessage);
+    return context.getString(action, (Object) actionArgs);
+  }
+
+  private static String getErrorSpecificMessage(Context context, Throwable error) {
+    if (error instanceof GenericException) {
+      GenericException exception = (GenericException) error;
+      return context.getString(
+          exception.getUserMessage(), (Object) exception.getUserMessageArguments());
+    } else if (error instanceof TimeoutException) {
+      return context.getString(R.string.timeout_exception);
+    } else {
+      // It is not a known error, let's log it but not show information to
+      // the average user as it will not be useful to him.
+      Log.d(
+          ErrorUtils.class.getSimpleName(),
+          "Error " + error.getClass() + " is not a know error type.");
```

Comment:
This is the most critical case, yet it seems the least noisy.
-> During development, it might be useful to actually show a toast message along the lines of "Unexpected error occurred - check the logs" .
-> I also don't see this logging the full details of the error (i.e. stack trace) whereas this is where we need it the most.
-> It seems this is logged as debug, it should be logged with error or critical , since this is an exception we haven't caught or planned for.

Reply:
> During development, it might be useful to actually show a toast message along the lines of "Unexpected error occurred - check the logs"
Yes, I can do that

>  I also don't see this logging the full details of the error (i.e. stack trace) whereas this is where we need it the most.
The error is always logged, this part is only for a nice message explaining the error. (Mostly for toast, but also as a log message) It is logged as part of line 39.

> It seems this is logged as debug, it should be logged with error or critical , since this is an exception we haven't caught or planned for.
Should this be logged in a production system ? This feels like debug only to me.

Reply:
I'll approve it to allow for the merge, but I would suggest adding the toast